src/Autofac/Core/Resolving/Pipeline/IPipelineResolveOperation.cs

```diff
@@ -0,0 +1,73 @@
+﻿// This software is part of the Autofac IoC container
+// Copyright © 2020 Autofac Contributors
+// https://autofac.org
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+using System.Collections.Generic;
+using Autofac.Core.Diagnostics;
+using Autofac.Core.Resolving.Middleware;
+
+namespace Autofac.Core.Resolving.Pipeline
+{
+    public interface IPipelineResolveOperation : IResolveOperation, IComponentContext, ITracingIdentifer
```

Comment:
The only class that would derive from it would be the ResolveOperation class, but yes, and abstract class may make more sense; I get what you're saying about the events. Happy to change that if you think it's semantically clearer.

Reply:
I think if there's a specific/prescribed order and timing to when the events need to be raised then likely it should be some built-in mechanism to ensure later derived stuff (if there is any) would always have the right order and timing.

Reply:
Ok, will swap this for an abstract class tomorrow.

Reply:
This is done. `IPipelineResolveOperation` is now `ResolveOperationBase`.  Calls to an abstract `ExecuteRequest` to allow the implementation to anything interesting, but most of the operation logic is in the base class.